tap_zendesk/streams.py

```diff
@@ -139,88 +141,110 @@ def sync(self, state):
+            should_yield = self._buffer_record((self.stream, ticket_dict))
 
-            should_yield = self._buffer_record(ticket_buffer, (self.stream, ticket_dict))
-            if should_yield:
-                for t in ticket_buffer:
-                    yield t
-                ticket_buffer = []
-
-            if audits_stream.is_selected():
-                if ticket_dict["status"] == "deleted":
-                    LOGGER.warning("Unable to retrieve audits for deleted ticket (ID: %s), skipping...", ticket_dict["id"])
-                    continue
+            if is_deleted:
+                LOGGER.warning("Unable to retrieve audits/metrics for deleted ticket (ID: %s), skipping...", ticket_dict["id"])
+
+            if audits_stream.is_selected() and not is_deleted:
                 for audit in audits_stream.sync(ticket_dict["id"]):
-                    should_yield = self._buffer_record(audit_buffer, audit)
-                    if should_yield:
-                        for a in audit_buffer:
-                            yield a
-                        audit_buffer = []
-                        emit_audits_metric()
-        if ticket_buffer:
-            for t in ticket_buffer:
-                yield t
-        if audit_buffer:
-            for a in audit_buffer:
-                yield a
-        emit_audits_metric()
+                    self._buffer_record(audit)
+
+            if metrics_stream.is_selected() and not is_deleted:
+                for metric in metrics_stream.sync(ticket_dict["id"]):
+                    self._buffer_record(metric)
+
+            if should_yield:
+                for rec in self._empty_buffer():
+                    yield rec
+                emit_sub_stream_metrics(audits_stream)
+                emit_sub_stream_metrics(metrics_stream)
+                singer.write_state(state)
+
+        for rec in self._empty_buffer():
+            yield rec
+        emit_sub_stream_metrics(audits_stream)
+        emit_sub_stream_metrics(metrics_stream)
+        singer.write_state(state)
 
 class TicketAudits(Stream):
     name = "ticket_audits"
     replication_method = "INCREMENTAL"
-    oldest_date = utils.strptime_with_tz('2018-07-06T18:01:35Z')
     count = 0
 
     def sync(self, ticket_id):
         ticket_audits = self.client.tickets.audits(ticket=ticket_id)
         for ticket_audit in ticket_audits:
             ticket_audit_dict = ticket_audit.to_dict()
             self.count += 1
-
-            if utils.strptime_with_tz(ticket_audit_dict["created_at"]) < self.oldest_date:
-                self.oldest_date = utils.strptime_with_tz(ticket_audit_dict["created_at"])
             yield (self.stream, ticket_audit_dict)
 
+class TicketMetrics(Stream):
+    name = "ticket_metrics"
+    replication_method = "INCREMENTAL"
+    count = 0
+
+    def sync(self, ticket_id):
+        ticket_metric = self.client.tickets.metrics(ticket=ticket_id)
+        ticket_metric_dict = ticket_metric.to_dict()
```

Comment:
The `to_dict` isn't actually necessary, not sure why we were doing it for Ticket, but this should be handled by the custom JSON Serializer in this tap.